packages/astro/src/vite-plugin-astro/hmr.ts

```diff
@@ -17,32 +22,24 @@ export interface HandleHotUpdateOptions {
 	config: AstroConfig;
 	logging: LogOptions;
 	compile: () => ReturnType<typeof cachedCompilation>;
+	source: string;
 }
 
 export async function handleHotUpdate(
 	ctx: HmrContext,
-	{ config, logging, compile }: HandleHotUpdateOptions
+	{ config, logging, compile, source }: HandleHotUpdateOptions
 ) {
 	let isStyleOnlyChange = false;
 	if (ctx.file.endsWith('.astro') && isCached(config, ctx.file)) {
 		// Get the compiled result from the cache
 		const oldResult = await compile();
-		// But we also need a fresh, uncached result to compare it to
+		// Skip HMR if source isn't changed
+		if (oldResult.source === source) return [];
+		// Invalidate to get fresh, uncached result to compare it to
 		invalidateCompilation(config, ctx.file);
 		const newResult = await compile();
-		// If the hashes are identical, we assume only styles have changed
-		if (oldResult.scope === newResult.scope) {
+		if (isStyleOnlyChanged(oldResult, newResult)) {
 			isStyleOnlyChange = true;
-			// All styles are the same, we can skip an HMR update
-			const styles = new Set(newResult.css);
-			for (const style of oldResult.css) {
-				if (styles.has(style)) {
-					styles.delete(style);
-				}
-			}
-			if (styles.size === 0) {
-				return [];
-			}
```

Comment:
Since scope hash now derives from `normalizedFilename` only, I created a `isStyleOnlyChanged` function to handle it. It's a bit tedious than before, but I think leaving the HMR logic out of the compiler could be nice for now.

I also remove the `styles.size === 0` handling. The logic isn't quite right given a scenario where:

1. `oldResult` has two `<style>`s
2. User removes one `<style>`
3. `newResult` now has one `<style>`
4. This logic makes `styles.size === 0` and skips HMR.
5. BUG: The removed style's side effect isn't removed as no HMR or reload is done.

It seems that this does handle skipping HMR if no source code change, which I handled on line 37 instead.